File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
vector<vector <int>> res;
void solve(int i,int target,vector<int> &arr,vector<vector <int>> &res,vector<int> & temp){
    if(i==arr.size()){
        if(target == 0){
            res.push_back(temp);
    

        }
        return;
    }
    if(arr[i] <= target){
        temp.push_back(arr[i]);
        solve(i,target-arr[i],arr,res,temp);
        temp.pop_back();
    }
    solve(i+1,target,arr,res,temp);
}
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        solve(0,target,candidates,res,temp);
        return res;
       
        

            
        
        
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    
    public void solve(int i,int target ,int arr[],List<Integer> temp,List<List<Integer>> res){
        if(arr.length==i){
            if(target==0){
                res.add(new ArrayList<>(temp));
            }
            return;
        }
        if(target>=arr[i]){
        temp.add(arr[i]);
        solve(i,target-arr[i],arr,temp,res);
        temp.remove(temp.size()-1);
    }
      solve(i + 1,target,arr,temp,res);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> temp=new ArrayList<>();
          List<List<Integer>> res=new ArrayList<>();
        Arrays.sort(candidates);
       

        solve(0,target,candidates,temp,res);
        return res;
        
        
    }
}